torch/lib/c10d/ProcessGroupNCCL.cpp

```diff
@@ -644,6 +644,13 @@ std::vector<std::shared_ptr<NCCLComm>>& ProcessGroupNCCL::getNCCLComm(
 
     // Creates the NCCL streams
     streamVal.push_back(at::cuda::getStreamFromPool());
+
+    // If single-process single-device mode, WorkNCCL::getFuture is supported,
+    // so get a dedicated stream to run FutureNCCL then callbacks.
+    if (devices.size() == 1) {
+      futureNCCLCallbackStream_ =
+          std::make_shared<at::cuda::CUDAStream>(at::cuda::getStreamFromPool());
+    }
```

Comment:
This should be done in the constructor of ProcessGroupNCCL and not each time we call `getNCCLComm`.

Reply:
I've decided to make a static vector of streams called `futureNCCLCallbackStreams_` instead that will store one stream per device. I think it will look way better. I'll give a quick try and let you know if it causes problems.

Reply:
That's a bit strange but when I try to initialize even a simple int using c10::cuda::device_count: `int ProcessGroupNCCL::test = c10::cuda::device_count();`; I get the following error:
```
THCudaCheck FAIL file=../aten/src/THC/THCGeneral.cpp line=54 error=3 : initialization error
ERROR:root:Caught exception: 
Traceback (most recent call last):
  File "/home/sinannasir/local/pytorch/torch/testing/_internal/common_distributed.py", line 226, in wrapper
    fn()
  File "/home/sinannasir/local/pytorch/torch/testing/_internal/common_distributed.py", line 93, in wrapper
    return func(*args, **kwargs)
  File "test/distributed/test_c10d.py", line 3226, in test_ddp_comm_hook_allreduce_with_then_hook_nccl
    gpu_model = self._gpu_model_with_ddp_comm_hook(
  File "test/distributed/test_c10d.py", line 3111, in _gpu_model_with_ddp_comm_hook
    TestDdpCommHook().to(device_id),
  File "/home/sinannasir/local/pytorch/torch/nn/modules/module.py", line 612, in to
    return self._apply(convert)
  File "/home/sinannasir/local/pytorch/torch/nn/modules/module.py", line 359, in _apply
    module._apply(fn)
  File "/home/sinannasir/local/pytorch/torch/nn/modules/module.py", line 381, in _apply
    param_applied = fn(param)
  File "/home/sinannasir/local/pytorch/torch/nn/modules/module.py", line 610, in convert
    return t.to(device, dtype if t.is_floating_point() else None, non_blocking)
  File "/home/sinannasir/local/pytorch/torch/cuda/__init__.py", line 173, in _lazy_init
    torch._C._cuda_init()
RuntimeError: cuda runtime error (3) : initialization error at ../aten/src/THC/THCGeneral.cpp:54
exiting process with exit code: 10
THCudaCheck FAIL file=../aten/src/THC/THCGeneral.cpp line=54 error=3 : initialization error
ERROR:root:Caught exception: 
Traceback (most recent call last):
  File "/home/sinannasir/local/pytorch/torch/testing/_internal/common_distributed.py", line 226, in wrapper
    fn()
  File "/home/sinannasir/local/pytorch/torch/testing/_internal/common_distributed.py", line 93, in wrapper
    return func(*args, **kwargs)
  File "test/distributed/test_c10d.py", line 3226, in test_ddp_comm_hook_allreduce_with_then_hook_nccl
    gpu_model = self._gpu_model_with_ddp_comm_hook(
  File "test/distributed/test_c10d.py", line 3111, in _gpu_model_with_ddp_comm_hook
    TestDdpCommHook().to(device_id),
  File "/home/sinannasir/local/pytorch/torch/nn/modules/module.py", line 612, in to
    return self._apply(convert)
  File "/home/sinannasir/local/pytorch/torch/nn/modules/module.py", line 359, in _apply
    module._apply(fn)
  File "/home/sinannasir/local/pytorch/torch/nn/modules/module.py", line 381, in _apply
    param_applied = fn(param)
  File "/home/sinannasir/local/pytorch/torch/nn/modules/module.py", line 610, in convert
    return t.to(device, dtype if t.is_floating_point() else None, non_blocking)
  File "/home/sinannasir/local/pytorch/torch/cuda/__init__.py", line 173, in _lazy_init
    torch._C._cuda_init()
RuntimeError: cuda runtime error (3) : initialization error at ../aten/src/THC/THCGeneral.cpp:54
```

I just initialize the streams inside constructor and they are not static.